File: Java/ReorderingTheCows.java
import java.io.*;

public class ReorderingTheCows {

    //static BufferedReader r = new BufferedReader(new InputStreamReader(System.in));
    //static PrintWriter pw = new PrintWriter(System.out);

    static int[] loc;
    public static void main(String[] args) throws IOException {
		BufferedReader r = new BufferedReader(new FileReader("reorder.in"));
		PrintWriter pw = new PrintWriter(new FileWriter("reorder.out"));

        int N = Integer.parseInt(r.readLine());

        int[] start = new int[N];
        for (int i = 0; i < N; i ++) {
            start[i] = Integer.parseInt(r.readLine()) - 1;
        }

        loc = new int[N];
        int[] goal = new int[N];
        for (int i = 0; i < N; i ++) {
            goal[i] = Integer.parseInt(r.readLine()) - 1;
            loc[goal[i]] = i;
        }

        int longest = -1;
        int shifts = 0;
        while (true) {
            boolean found = false;
            for (int i = 0; i < N; i ++) {
                if (start[i] != goal[i]) {
                    longest = Math.max(longest,shift(start, i));
                    shifts ++;
                    found = true;
                }
            }
            if (!found) {
                break;
            }
        }

        pw.println(shifts + " " + longest);
        pw.close();
    }
    public static int shift(int[] arr, int start) {
        int counter = 0;

        int val = arr[start];
        int index = loc[val];

        while (true) {
            int nextVal = arr[index];
            int nextIndex = loc[nextVal];

            arr[index] = val;
            counter ++;
            if (index == start) {
                break;
            }

            val = nextVal;
            index = nextIndex;
        }
        return counter;
    }
}

File: Python/Reordering_The_Cows.py
import sys

def cycle(start):
    counter = 0
    d = start
    stop = False
    
    while not stop:
        counter += 1
        visited[d] = True
        
        d = loc[a[d]]
        if d == start:
            stop = True
           
    return counter

sys.stdin = open("reorder.in","r")
sys.stdout = open("reorder.out","w")
n = int(input())

a = []
for i in range(n):
    a.append(int(input()))

b = []

for i in range(n):
    b.append(int(input()))
    
loc = [0] * (n+1)

for i in range(n):
    loc[b[i]] = i

cycles = 0; longest = 0

original = a.copy()
visited = [False] * n
for i in range(n):
    if a[i] != b[i] and not visited[i]:
        length = cycle(i)
        longest = max(longest,length)
        cycles += 1

if cycles == 0:
    print(cycles,-1)
else:
    print(cycles,longest)